### PR TITLE
fix(core): nullable `array` types in `3.1.x` specs

### PR DESCRIPTION
**Pull request by joscha**
Currently, in 3.1 specs, the follwing:

```yaml
      properties:
        arrayDataOrNull:
          items:
            $ref: "#/components/schemas/SomeObject"
          type:
            - array
            - 'null'
```
yields:
```ts
'arrayDataOrNull': Array<SomeObject>
```
where it should yield:
```ts
'arrayDataOrNull': Array<SomeObject> | null;
```

It seems either nullability is not determined correctly in core and/or the `typescript` generator doesn't interpret it correctly.

Real-life example: https://github.com/planet-a-ventures/affinity-node/pull/49
Discussion: https://openapi-generator.slack.com/archives/CLSB0U0R5/p1727342905017529

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming `7.x.0` minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [ ] ~If your PR is targeting a particular programming language~ I think this is a core issue: [@wing328](https://github.com/wing328) (2015/07) [❤️](https://www.patreon.com/wing328)
[@jimschubert](https://github.com/jimschubert) (2016/05) [❤️](https://www.patreon.com/jimschubert)
[@cbornet](https://github.com/cbornet) (2016/05)
[@jmini](https://github.com/jmini) (2018/04) [❤️](https://www.patreon.com/jmini)
[@etherealjoy](https://github.com/etherealjoy) (2019/06)

### CODE REVIEW

File: modules/openapi-generator/src/test/resources/3_1/null-types-simple.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  schemas:
+    WithNullableType:
+      type: object
+      required:
+        # This simplifies the output
+        - arrayDataOrNull
+        - stringDataOrNull
+        - oneofOrNull
+      properties:
+        arrayDataOrNull:
```
Comment (joscha) [inline modules/openapi-generator/src/test/resources/3_1/null-types-simple.yaml:15]: This is the broken part

File: samples/client/others/typescript/builds/null-types-simple/models/WithNullableType.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
+ * https://openapi-generator.tech
+ * Do not edit the class manually.
+ */
+
+import { SomeObject } from '../models/SomeObject';
+import { HttpFile } from '../http/http';
+
+export class WithNullableType {
+    'arrayDataOrNull': Array<SomeObject>;
```
Comment (joscha) [inline samples/client/others/typescript/builds/null-types-simple/models/WithNullableType.ts:17]: Should be:
```suggestion
    'arrayDataOrNull': Array<SomeObject> | null;
```

File: samples/client/others/typescript/builds/null-types-simple/models/WithNullableType.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
+ * https://openapi-generator.tech
+ * Do not edit the class manually.
+ */
+
+import { SomeObject } from '../models/SomeObject';
+import { HttpFile } from '../http/http';
+
+export class WithNullableType {
+    'arrayDataOrNull': Array<SomeObject> | null;
```
Comment (joscha) [inline samples/client/others/typescript/builds/null-types-simple/models/WithNullableType.ts:17]: This line is the fix - it now has the `| null`

File: samples/client/others/typescript/.gitattributes [hunk 4]
```diff
@@ -0,0 +1 @@
+builds/**/* linguist-generated=true
```
Comment (joscha) [inline samples/client/others/typescript/.gitattributes:1]: This makes sure that all the generated code collapses in the github diff.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/OpenAPINormalizer.java [hunk 5]
```diff
@@ -1268,6 +1268,7 @@ private Schema processNormalize31Spec(Schema schema, Set<Schema> visitedSchemas)
                 as.setMaxItems(schema.getMaxItems());
                 as.setExtensions(schema.getExtensions());
                 as.setXml(schema.getXml());
+                as.setNullable(schema.getNullable());
```
Comment (joscha) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/OpenAPINormalizer.java:1271]: Further up, we fix up the `nullable` flag when the types contain the new 3.1 `null` type: https://github.com/OpenAPITools/openapi-generator/blob/29c3b020e620330669bac9950863b09bc2e61764/modules/openapi-generator/src/main/java/org/openapitools/codegen/OpenAPINormalizer.java#L1249-L1252

but then, if there's a single type left AND it is an array schema, we forget to copy this property to the cloned `ArraySchema`. This line makes sure that the nullable flag is transferred onto the schema clone.

Comment (joscha) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/OpenAPINormalizer.java:1271]: This issue is likely to happen again in the future when/if a new property is added that is in 3.1+ but not 3.0.x - we could possibly introduce a `clone` or `transfer` function somewhere which would make this a bit more obvious, but it's not a guarantee either.

**Review by macjohnny — APPROVED** [review]
thanks!

**Comment by wing328** [comment]
lgtm. thanks for the fix.

i will file another PR to add a unit test later this week

**Comment by joscha** [comment]
> i will file another PR to add a unit test later this week

Quickly added one: https://github.com/OpenAPITools/openapi-generator/pull/19714

**Pull request merged**